src/parser/Parser.ts

```diff
@@ -24,797 +26,872 @@ type BlockTerminator =
     Lexeme.EndSub |
     Lexeme.EndFunction;
 
-let current: number;
-let tokens: ReadonlyArray<Token>;
+/** The results of a Parser's parsing pass. */
+interface ParseResults {
+    /** The statements produced by the parser. */
+    statements: ReadonlyArray<Stmt.Statement>,
+    /** The errors encountered by the Parser. */
+    errors: ReadonlyArray<ParseError>
+}
 
-export function parse(toParse: ReadonlyArray<Token>) {
-    current = 0;
-    tokens = toParse;
+export class Parser {
+    /** Allows consumers to observe errors as they're detected. */
+    readonly events = new EventEmitter();
 
-    let statements: Statement[] = [];
+    /**
+     * Parses an array of `Token`s into an abstract syntax tree that can be executed with the `Interpreter`.
+     * @param toParse the array of tokens to parse
+     * @returns an array of `Statement` objects that together form the abstract syntax tree of the program
+     */
+    parse(toParse: ReadonlyArray<Token>): ParseResults {
```

Comment:
You have a static `scan()` convenience method on the lexer, but you didn't include a static `parse()` convenience method for the parser. Was this intentional?